api/v1alpha3/virtualmachine_types.go

```diff
@@ -208,28 +208,53 @@ const (
 	VirtualMachinePowerOpModeTrySoft VirtualMachinePowerOpMode = "TrySoft"
 )
 
+type VirtualMachineImageRef struct {
+	// Kind describes the type of image, either a namespace-scoped
+	// VirtualMachineImage or cluster-scoped ClusterVirtualMachineImage.
+	Kind string `json:"kind"`
+
+	// Name refers to the name of a VirtualMachineImage resource in the same
+	// namespace as this VM or a cluster-scoped ClusterVirtualMachineImage.
+	Name string `json:"name"`
+}
+
 // VirtualMachineSpec defines the desired state of a VirtualMachine.
 type VirtualMachineSpec struct {
+
+	// +optional
+
+	// Image describes the reference to the VirtualMachineImage or
+	// ClusterVirtualMachineImage resource used to deploy this VM.
+	//
+	// Please note, unlike the field spec.imageName, the value of
+	// spec.image.name MUST be a Kubernetes object name.
+	//
+	// Please also note, when this field is non-empty, the value of the
+	// field spec.imageName is ignored.
+	Image *VirtualMachineImageRef `json:"image,omitempty"`
+
+	// +optional
+
 	// ImageName describes the name of the image resource used to deploy this
 	// VM.
 	//
 	// This field may be used to specify the name of a VirtualMachineImage
 	// or ClusterVirtualMachineImage resource. The resolver first checks to see
-	// if there is a VirtualMachineImage with the specified name. If no
-	// such resource exists, the resolver then checks to see if there is a
-	// ClusterVirtualMachineImage resource with the specified name in the same
-	// Namespace as the VM being deployed.
+	// if there is a VirtualMachineImage with the specified name in the
+	// same namespace as the VM being deployed. If no such resource exists, the
+	// resolver then checks to see if there is a ClusterVirtualMachineImage
+	// resource with the specified name.
 	//
 	// This field may also be used to specify the display name (vSphere name) of
 	// a VirtualMachineImage or ClusterVirtualMachineImage resource. If the
 	// display name unambiguously resolves to a distinct VM image (among all
 	// existing VirtualMachineImages in the VM's namespace and all existing
-	// ClusterVirtualMachineImages), then a mutation webhook updates this field
-	// with the VM image resource name. If the display name resolves to multiple
-	// or no VM images, then the mutation webhook denies the request and outputs
-	// an error message accordingly.
+	// ClusterVirtualMachineImages), then a mutation webhook updates the
+	// spec.image field with the reference to the resolved VM image. If the
```

Comment:
Does this mean that clients should ignore this field after the spec is mutated? I can imagine the following 3 cases:

1. user specifies spec.image, and the match spec.imageName
2. user specifies spec.image, and a mis-match spec.imageName
3. user only specified this spec.imageName

For the first 2 cases, the spec.imageName is ignored, for the 3rd one, the web hook updates the spec.image based on the imageName. After the spec is mutated, case 1 and case 3 becomes the same, which is ok as it may not be necessary to track what is the original request. Case 2 can be a little confusing as it shows 2 mis-matching image in the spec (from UX perspective), it could be avoided with a check.

Reply:
cc @sreyasn since he asked a similar question up above

The reason we do not ensure `spec.image.name` and `spec.imageName` are the same on create is because the validation webhook runs after mutation, and so if `spec.imageName` is a display name, then `spec.image.name` will always be different, even if pointing to the same image.

The only way to do the above would be to do the validation during mutation.

Reply:
I think I see a way to do this. Will update the PR.

Reply:
Okay, this is addressed.